### ISSUE TITLE
[Due for payment 2025-01-18] [$250] Notification – Have no sound for notification received

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#53700]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.72-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** testpayment935+775@gmail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in as user A
3. Open chat with user B
4. As user B send message to user A
5. Notice that user A receive a notification

## Expected Result:
Notification is with sound

## Actual Result:
Notification is without sound

## Workaround:
Unknown

## Platforms:
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

https://github.com/user-attachments/assets/81254557-0a12-412a-96f8-0300b73a586e

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021866271667566111224</li>
        <li>Upwork Job ID: 1866271667566111224</li>
        <li>Last Price Increase: 2024-12-10</li>
<li>Automatic offers: </li>
<ul>
<li>truph01 | Contributor | 105599028</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @RachCHopkins</details>

**Comment by RachCHopkins** [comment]
Can repro!

**Comment by JoshIri360** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue
Browser notifications for new messages in New Expensify are currently playing without sound, when they should play a notification sound to alert users of new messages.

### What is the root cause of that problem?
The root cause is twofold:
1. The `showCommentNotification` method was setting `silent: true` by default
2. There was no sound playback implementation even when notifications weren't silent

### What changes do you think we should make in order to solve the problem?
We need to make two key changes:

1. In `BrowserNotifications.ts`, change the `showCommentNotification` to use `silent: false`:
```typescript:src/libs/Notification/LocalNotification/BrowserNotifications.ts
showCommentNotification(report: Report, reportAction: ReportAction, onClick: LocalNotificationClickHandler) {
    // ... existing code ...
    push(title, body, icon, data, onClick, false);
},
```

2. Add sound playback in the `push` function:
```typescript:src/libs/Notification/LocalNotification/BrowserNotifications.ts
function push(
    title: string,
    body: string,
    icon: string,
    data: LocalNotificationData,
    onClick: LocalNotificationClickHandler,
    silent: boolean,
) {
    // ... existing code ...

    // Play notification sound if not silent
    if (!silent) {
        playSound(SOUNDS.RECEIVE);
    }

    // Continue with notification creation
    const notificationID = Str.guid();
    // ... rest of existing code ...
}
```

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?
N/A

### What alternative solutions did you explore? (Optional)
N/A

### Result
https://github.com/user-attachments/assets/d841fe52-51f3-4e8c-b6da-16c673baab1a

**Comment by jayeshmangwani** [comment]
Not overdue, testing the above 2 proposals

**Comment by jayeshmangwani** [comment]
@JoshIri360  and @truph01 , did you receive the notification sound on desktop in the current version?

**Comment by JoshIri360** [comment]
No I didn't, tested in chrome and safari.

**Comment by truph01** [comment]
> So we don’t get sound on the web (Chrome + Safari) or desktop either, right? Just confirming so I can test solutions on those platforms.

No. I cannot get the sound on Macos Chrome. With Safari, it works well.

**Comment by truph01** [comment]
@jayeshmangwani In Macos Safari, it works well. The issue only appears in Macos Chrome. But you need to hard-code the 6th param to ```false```:

https://github.com/Expensify/App/blob/082e14b9bc1259c329c7cd186ee8f21c0397f6fb/src/libs/Notification/LocalNotification/BrowserNotifications.ts#L125

before testing the sound notification if you want the sound is triggered by sending a message.

**Comment by truph01** [comment]
> it works fine on Safari, Chrome, and desktop.

In safari, that solution can introduce the bug when the sound is triggered 2 times. Did you encounter it? @jayeshmangwani

**Comment by jayeshmangwani** [comment]
No @truph01 , I have tested it a few times on Safari and only received one notification sound.

**Comment by truph01** [comment]
>  I have tested it a few times on Safari and only received one notification sound.

That means [this bug](https://github.com/Expensify/App/issues/53700) appears in both Chrome and Safari from your side, right?

**Comment by jayeshmangwani** [comment]
Yes, that’s correct, which is why I asked for [confirmation earlier](https://github.com/Expensify/App/issues/53700#issuecomment-2541299506). For me, neither Safari nor Chrome is producing any sound.

**Comment by truph01** [comment]
Here is the video from my side, the sound is triggered in Safari:

https://github.com/user-attachments/assets/8cecc9f9-fbc5-4721-acb0-c21d9195e59b

**Comment by truph01** [comment]
Can you share the notification setting from your device, please? Here is mine:

![image](https://github.com/user-attachments/assets/e17a1331-c2da-43b7-8105-9a6d005d5be1)

**Comment by jayeshmangwani** [comment]
<img width="505" alt="Screenshot 2024-12-14 at 12 09 33" src="https://github.com/user-attachments/assets/4eb44912-d8fa-460c-a948-b40e6cfa2dea" />

**Comment by jayeshmangwani** [comment]
> Here is the video from my side, the sound is triggered in Safari:

Interesting!

**Comment by truph01** [comment]
Also, can you give me your safari version, please? Based on the docs, the silent prop only works with Safari version > 16.6

**Comment by jayeshmangwani** [comment]
Version 17.2.1 (19617.1.17.11.12)

**Comment by truph01** [comment]
I tested in [another app](https://push.foo) and it still works in my Safari. In the video below, I tested with both values of prop ```silent```:




https://github.com/user-attachments/assets/7e3b1231-43e1-40dc-8637-5e190212dd6b


@jayeshmangwani Please help try it once you have a chance. Thanks

**Comment by jayeshmangwani** [comment]
@truph01 , sorry, but I’m not entirely clear from the video. When I try the app on my end using Safari, I hear sound when Silent Mode is disabled, but there’s no sound when Silent Mode is enabled.

**Comment by truph01** [comment]
> I hear sound when Silent Mode is disabled, but there’s no sound when Silent Mode is enabled

I think it is expected. When ```silent: true```, that means no sound should be triggered and vice versa. Please correct me if i was wrong.

**Comment by jayeshmangwani** [comment]
Yes, that’s definitely the expected behavior. However, what I’m trying to understand is: what’s the purpose of [this video](https://github.com/Expensify/App/issues/53700#issuecomment-2543130469)?

**Comment by truph01** [comment]
Ah, I included the video to demonstrate that the ```silent``` option is working correctly in Safari.

**Comment by jayeshmangwani** [comment]
Thanks, @truph01, for the details, but @JoshIri360's Proposal works well for me, and IMO we can go with it.

**Comment by jayeshmangwani** [comment]
@JoshIri360's [Proposal](https://github.com/Expensify/App/issues/53700#issuecomment-2537406732) works well for me and looks good, we can add custom playSound for web and desktop

🎀 👀 🎀 C+ reviewed

**Comment by jayeshmangwani** [comment]
For that we can add custom playSound 


```
    // Play notification sound if not silent
    if (!silent) {
        playSound(SOUNDS.RECEIVE);
    }

```

**Comment by jayeshmangwani** [comment]
@truph01 From the issue description's checklist, I can also see that tested marked the desktop and Chrome/Safari platforms where this issue is occuring .

**Comment by jayeshmangwani** [comment]
Sure thing, here it is
Uploading notificaiton-sound.mov…

**Comment by MarioExpensify** [comment]
Heey! Let's move forward with @JoshIri360's proposal, thanks everybody!

**Comment by JoshIri360** [comment]
I have applied to the job on Upwork and target to have the PR ready in a few hours

**Comment by JoshIri360** [comment]
PR [here](https://github.com/Expensify/App/pull/54315)

**Comment by jayeshmangwani** [comment]
@truph01 Could you please check out the PR branch and test it ? Did you hear two sounds?

In the video, you mentioned that the first sound is custom, which plays when we send a message (not the notification). The second sound is from the desktop notification.

Can you clarify the sequence you're experiencing? Did you hear the custom notification sound first, followed by the default notification sound, or the other way around?

**Comment by jayeshmangwani** [comment]
> After removing it, you will notice that the notification is received with the default desktop notification sound.

Yes, upon commenting out this line, I’m getting the default desktop notification sound, but no sound from Chrome.

**Comment by jayeshmangwani** [comment]
However, I’m still unable to hear two sounds from the desktop when this line is uncommented 🤷.

**Comment by jayeshmangwani** [comment]
Here’s how I am testing it—please let me know if I’m doing something wrong.


https://github.com/user-attachments/assets/9dc7dbfc-072a-4d3e-b1c8-8b87130403fe

**Comment by JoshIri360** [comment]
@truph01 @jayeshmangwani  will adding `silent: true` to the new Notification solve this.

```
notificationCache[notificationID] = new Notification(title, {
        body,
        icon: String(icon),
        data,
        silent: true,
        tag,
});
```

If this is a bug from the Chrome web browser and it's fixed, we could expect double notifications but with this, only our custom sound via playSound(SOUNDS.RECEIVE) will play and the default browser notification sound will be suppressed.

**Comment by JoshIri360** [comment]
@truph01 No, I don't but I think this may be because I can't seem to enable notification banners and sounds for the Desktop platform.

**Comment by JoshIri360** [comment]
@truph01 Yes, 2 notifications here

**Comment by JoshIri360** [comment]
[This](https://github.com/Expensify/App/issues/53700#issuecomment-2559085092) solves it, no?

**Comment by jayeshmangwani** [comment]
> Please ensure that after applying the code changes, you refresh the desktop app to ensure the changes are properly applied.

Yes, I’ve made sure to reload the desktop, but strangely, you and @JoshIri360 consistently get two sounds. I’m probably doing something wrong in my testing 😄

**Comment by JoshIri360** [comment]
You could try turning on DND, it would silence the system sound so you can listen for the custom one better.

**Comment by jayeshmangwani** [comment]
> You could try turning on DND, it would silence the system sound so you can listen for the custom one better.

Does enabling DND also not stop notifications?

**Comment by JoshIri360** [comment]
> Does enabling DND also not stop notifications?

No, only stops the banners and system sounds

**Comment by jayeshmangwani** [comment]
@MarioExpensify ,[Previously selected](https://github.com/Expensify/App/issues/53700#issuecomment-2537406732) proposal resulted in [two sounds](https://github.com/Expensify/App/issues/53700#issuecomment-2559020454). We need to revisit and rethink possible solutions for this scenario to ensure the expected behavior. Below are the key points outlined:

When constructing `Notification`, we are currently passing `silent: true`, which results in no sound for desktop and web notifications on staging. 
https://github.com/Expensify/App/blob/8b7096fc4c4d522407a6f51b4fe0fd3543e5feb2/src/libs/Notification/LocalNotification/BrowserNotifications.ts#L64-L68

**Possible Solutions:**
1. Pass `silent: false`:

    - This resolves the issue for desktop and Safari notifications.
   - However, there is a known Chrome bug where notifications remain silent despite `silent: false`. This seems to be an internal Chrome issue and has been observed on multiple sites.
   1. https://support.google.com/chrome/thread/269928241/sound-stopped-working-in-chrome-mac?hl=en
   2. https://discussions.apple.com/thread/254098862?sortBy=rank
    - We could accept this behavior on Chrome for now, assuming it may eventually be resolved by Chrome, **OR** we could implement a custom sound specifically for Chrome to address this limitation.
2. Use a custom sound for all platforms:

   - Playing a custom sound across all platforms ensures consistency.
   - Adding `playsound(SOUNDS.RECEIVE)` can provide a uniform experience on desktop, Safari, and Chrome.

**Comment by jayeshmangwani** [comment]
@truph01 your above solutions looks good and will def solve the problem at hand, but I think we should first reach a consensus on the expected result. Personally, I’m more inclined towards using only silent: false, since the Chrome issue might eventually be resolved on their end.

Alternatively, if we want a more immediate and consistent solution, we could consider using a custom sound across all platforms.

**Comment by JoshIri360** [comment]
@jayeshmangwani If we're using a custom sound across all platforms, [this](https://github.com/Expensify/App/issues/53700#issuecomment-2559085092) would be a viable solution.

This would silence the system notifications entirely, the major drawback would be the sound coming through when DND (same for web apps that use custom sounds for Notification like Discord and Whatsapp on web) is turned on as there isn't any support to check for this on Web yet.

**Comment by jayeshmangwani** [comment]
> If we're using a custom sound across all platforms, https://github.com/Expensify/App/issues/53700#issuecomment-2559085092 would be a viable solution.

Let’s wait for @MarioExpensify to reach a conclusion on whether we want to use custom sounds for all platforms.

**Comment by MarioExpensify** [comment]
Hello @jayeshmangwani, thanks for bringing this up to discussion. I don't like the idea that we would either accept Chrome's fault or have a custom code path only to address their issue.

I'm inclined towards the custom notification sound (and make everything consistent), but I'll take this discussion to product/design so we have their perspective on this as well. Please hold a little bit more on this before moving forward.

**Comment by arosiclair** [comment]
We usually handle and play notification sounds [here](https://github.com/Expensify/App/blob/b416e250cc633c0505390af6e567b0ccdeea16fd/src/libs/actions/User.ts#L840-L879). We implemented that in [this PR](https://github.com/Expensify/App/pull/31055). Why isn't that working?

**Comment by JoshIri360** [comment]
@arosiclair  The notification sounds only play when you're in that chat and the browser app is active.

https://github.com/user-attachments/assets/33211292-042d-4732-9537-69fe6978f055

**Comment by arosiclair** [comment]
> @arosiclair The notification sounds only play when you're in that chat and the browser app is active.

Thanks, that makes sense. So when we added `playSound` with push notifications in your original proposal, why did that cause duplicate sounds to play?

**Comment by arosiclair** [comment]
> The issue was that chrome notifications weren’t playing the default notification sound, adding the custom notification fixed that on Chrome, but for platforms where it already worked before, it added another sound. 

There was no sound because we set `silent` to `true` for comment notifications [here](https://github.com/Expensify/App/blob/4ffbd422296227ae003bfe0b5e584188bae5d68d/src/libs/Notification/LocalNotification/BrowserNotifications.ts#L125), right? So on desktop, the notification still plays a sound even with it set to silent?

**Comment by JoshIri360** [comment]
> There was no sound because we set `silent` to `true` for comment notifications [here](https://github.com/Expensify/App/blob/4ffbd422296227ae003bfe0b5e584188bae5d68d/src/libs/Notification/LocalNotification/BrowserNotifications.ts#L125), right? So on desktop, the notification still plays a sound even with it set to silent?

You're right, the silent: false change aimed to fix Chrome but caused duplicate sounds on platforms where sound already worked. On desktop, it seems the notification sound doesn't play with silent enabled.

The thought process was to give us more granular control over notification behavior across platforms. Keeping silent: true and adding playSound(SOUNDS.RECEIVE) during push should resolve this cleanly.

**Comment by arosiclair** [comment]
> The thought process was to give us more granular control over notification behavior across platforms. Keeping silent: true and adding playSound(SOUNDS.RECEIVE) during push should resolve this cleanly.

Cool, thanks for the explanation.

> Which approach do you prefer?

We're discussing this internally.

**Comment by MarioExpensify** [comment]
@jayeshmangwani after internally discussing we decided to move forward with the custom notification sound for all web-based platforms, we'll achieve a nice consistency (using the notification sound we already have when the app is opened) and should fix the Chrome issue.

**Comment by JoshIri360** [comment]
Thanks @MarioExpensify, @jayeshmangwani should I go ahead and edit my PR to match the intended behavior?

**Comment by jayeshmangwani** [comment]
@JoshIri360 Before making changes to the PR, could you please add a formal proposal for this new requirement as a comment here? @truph01, feel free to update the proposal if it hasn’t already been updated.

Additionally, I’m unsure how we should handle the compensation in this case @MarioExpensify, as the initially selected proposal has an issue that I overlooked. @truph01 helped identify the problem, and their contribution deserves acknowledgment. We can decide on the compensation after the PR is merged, potentially by splitting the payment from the mine. Just sharing my thoughts here.

**Comment by MarioExpensify** [comment]
@jayeshmangwani you've got a very good point, thanks for being so proactive about this! I'll reach out to @RachCHopkins and we'll circle back to you.

**Comment by jayeshmangwani** [comment]
Cool, @truph01's [Proposal](https://github.com/Expensify/App/issues/53700#issuecomment-2537853394) meets our new requirements. We will add `silent: true` and call `playSound` when `silent` prop value is `false`.

cc: @MarioExpensify

**Comment by JoshIri360** [comment]
> @JoshIri360 Before making changes to the PR, could you please add a formal proposal for this new requirement as a comment here? @truph01, feel free to update the proposal if it hasn’t already been updated.
> 
> Additionally, I’m unsure how we should handle the compensation in this case @MarioExpensify, as the initially selected proposal has an issue that I overlooked. @truph01 helped identify the problem, and their contribution deserves acknowledgment. We can decide on the compensation after the PR is merged, potentially by splitting the payment from the mine. Just sharing my thoughts here.



Updated [here](https://github.com/Expensify/App/issues/53700#issuecomment-2537406732)

**Comment by JoshIri360** [comment]
Sorry, updated [this](https://github.com/Expensify/App/issues/53700#issuecomment-2568413701), should be a comment.

**Comment by jayeshmangwani** [comment]
@truph01 Sorry, I didn’t quite understand what you’re saying here. However, I’ve already commented here https://github.com/Expensify/App/issues/53700#issuecomment-2568375793 that your proposal seems like it will solve this issue. 
Does this [proposal](https://github.com/Expensify/App/issues/53700#issuecomment-2537853394) have any specific concerns? I believe it should work perfectly for custom notification sounds since we’re suggesting passing `silent: true`, with the default value for `silent` being `false`. If `silent` is `false`, we can play the custom sound using `playSound`.

**Comment by jayeshmangwani** [comment]
Mario, let's reassign this issue to @truph01 since their [initial proposal](https://github.com/Expensify/App/issues/53700#issuecomment-2537853394) complies with our updated [expected results](https://github.com/Expensify/App/issues/53700#issuecomment-2568154962).
Thanks!

**Comment by MarioExpensify** [comment]
Thank you @jayeshmangwani, we'll move forward with @truph01's proposal then. We are discussing internally about how payment will be handled for this as it got a little off-script, I'll circle back when we have the final decision.

@truph01 feel free to start your PR. Thank you.

**Comment by jayeshmangwani** [comment]
# BugZero Checklist:

- [x] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [x] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [x] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [x] 3d. QA
  - [ ] 3z. Other:

</details>

- [x] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment: https://github.com/Expensify/App/pull/31055#discussion_r1921609424

- [x] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: N/A

- [x] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

### Regression Test Proposal

1. Log in as user A on Desktop or web
2. Open chat with user B
3. As user B sends message to user A
4. Verify that user A receives a notification with a sound.

Do we agree 👍 or 👎

**Comment by RachCHopkins** [comment]
Payment Summary:

- Contributor: @truph01 to be paid $250 via Upwork
- Contributor+: @jayeshmangwani to be paid $250 via NewDot manual request

Upwork job [here](https://www.upwork.com/jobs/~021866271667566111224)

**Comment by RachCHopkins** [comment]
Contributor has been paid, the contract has been completed, and the Upwork post has been closed.

**Comment by garrettmknight** [comment]
$250 approved for @jayeshmangwani
